# Name Trao An Huy
# MSSV: 22280041
# Lab 1 NLP ngram tf idf

# Dataset

## Download dataset
Vietnamese Students' Feedback Corpus (UIT-VSFC) is the resource consists of over 16,000 sentences which are human-annotated with two different tasks: sentiment-based and topic-based classifications.

[1] Kiet Van Nguyen, Vu Duc Nguyen, Phu Xuan-Vinh Nguyen, Tham Thi-Hong Truong, Ngan Luu-Thuy Nguyen, UIT-VSFC: Vietnamese Students' Feedback Corpus for Sentiment Analysis,  2018 10th International Conference on Knowledge and Systems Engineering (KSE 2018), November 1-3, 2018, Ho Chi Minh City, Vietnam

In [1]:
pip install datasets


   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 491.2/491.2 kB 6.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 116.3/116.3 kB 5.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 183.9/183.9 kB 8.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 143.5/143.5 kB 6.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 194.8/194.8 kB 6.7 MB/s eta 0:00:00
  Attempting uninstall: fsspec
    Found existing installation: fsspec 2025.3.0
    Uninstalling fsspec-2025.3.0:
      Successfully uninstalled fsspec-2025.3.0
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
gcsfs 2025.3.0 requires fsspec==2025.3.0, but you have fsspec 2024.12.0 which is incompatible.
torch 2.6.0+cu124 requires nvidia-cublas-cu12==12.4.5.8; platform_system == "Linux" and platform_machine == "x86_64", but you have nvidia-cublas-cu12 12.5.3.2 which is i

In [2]:
from datasets import load_dataset

dataset = load_dataset("uitnlp/vietnamese_students_feedback")

/usr/local/lib/python3.11/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


README.md:   0%|          | 0.00/4.79k [00:00<?, ?B/s]

vietnamese_students_feedback.py:   0%|          | 0.00/4.97k [00:00<?, ?B/s]

0000.parquet:   0%|          | 0.00/475k [00:00<?, ?B/s]

0000.parquet:   0%|          | 0.00/63.3k [00:00<?, ?B/s]

0000.parquet:   0%|          | 0.00/134k [00:00<?, ?B/s]

Generating train split:   0%|          | 0/11426 [00:00<?, ? examples/s]

Generating validation split:   0%|          | 0/1583 [00:00<?, ? examples/s]

Generating test split:   0%|          | 0/3166 [00:00<?, ? examples/s]

## Interacting with the downloaded data

In [3]:
train_set = dataset['train']
train_set

Dataset({
    features: ['sentence', 'sentiment', 'topic'],
    num_rows: 11426
})

In [4]:
train_set[0]

{'sentence': 'slide giáo trình đầy đủ .', 'sentiment': 2, 'topic': 1}

In [5]:
len(train_set)

11426

## Split a sentence

In [6]:
# Read a sentence
example_word_list = train_set[0]['sentence']
example_word_list

'slide giáo trình đầy đủ .'

In [7]:
# Split sentence word-by-word
example_word_list.split()

['slide', 'giáo', 'trình', 'đầy', 'đủ', '.']

In [8]:
# Join words into 1 full sentence
sentence = ""
for word in example_word_list:
    sentence += word
sentence

'slide giáo trình đầy đủ .'

In [9]:
# Get 10 sentences to process
sentence_list = []
for idx in range(10):
    sentence = ""
    for word in train_set[idx]['sentence']:
        sentence += word
    sentence_list.append(sentence)
sentence_list

['slide giáo trình đầy đủ .',
 'nhiệt tình giảng dạy , gần gũi với sinh viên .',
 'đi học đầy đủ full điểm chuyên cần .',
 'chưa áp dụng công nghệ thông tin và các thiết bị hỗ trợ cho việc giảng dạy .',
 'thầy giảng bài hay , có nhiều bài tập ví dụ ngay trên lớp .',
 'giảng viên đảm bảo thời gian lên lớp , tích cực trả lời câu hỏi của sinh viên , thường xuyên đặt câu hỏi cho sinh viên .',
 'em sẽ nợ môn này , nhưng em sẽ học lại ở các học kỳ kế tiếp .',
 'thời lượng học quá dài , không đảm bảo tiếp thu hiệu quả .',
 'nội dung môn học có phần thiếu trọng tâm , hầu như là chung chung , khái quát khiến sinh viên rất khó nắm được nội dung môn học .',
 'cần nói rõ hơn bằng cách trình bày lên bảng thay vì nhìn vào slide .']

# Text processing

## N-grams
- N-grams are continuous sequences of words or symbols, or tokens in a document. In technical terms, they can be defined as the neighboring sequences of items in a document.
- We can use n-grams or multiple other text preprocessing algorithms by incorporating [`nltk`](https://www.nltk.org/) library.

In [10]:
example_sentence = [
    'This is the first document.',
    'This document is the second document.',
    'And this is the third one.',
    'Is this the first document?',
]

In [11]:
pip install nltk

In [12]:
from nltk import ngrams
import numpy as np

num_of_grams = np.arange(1, 4, 1)  # Test 3 n-grams

print("Original sentence:", example_sentence[1])
print("===" * 5)

for gram in num_of_grams:
    splitted_sentence = ngrams(example_sentence[1].split(), int(gram))  # Convert gram to int
    print(f"{gram}-gram: ", end='')
    n_grams_list = [' '.join(grams) for grams in splitted_sentence]
    print(n_grams_list)
    print()

Original sentence: This document is the second document.
1-gram: ['This', 'document', 'is', 'the', 'second', 'document.']

2-gram: ['This document', 'document is', 'is the', 'the second', 'second document.']

3-gram: ['This document is', 'document is the', 'is the second', 'the second document.']



## Extract features with n-grams

In [13]:
import pandas as pd
# count the frequency of each n-gram
from sklearn.feature_extraction.text import CountVectorizer

In [14]:
count_vectorize_model = CountVectorizer(ngram_range = (1, 1))
n_grams_feature_vector = count_vectorize_model.fit_transform(example_sentence).toarray()
word_frequency = pd.DataFrame(data = n_grams_feature_vector, columns = count_vectorize_model.get_feature_names_out())
word_frequency.T

0  1  2  3
and       0  0  1  0
document  1  2  0  1
first     1  0  0  1
is        1  1  1  1
one       0  0  1  0
second    0  1  0  0
the       1  1  1  1
third     0  0  1  0
this      1  1  1  1

In [15]:
count_vectorize_model = CountVectorizer(ngram_range = (1, 1))

n_grams_feature_vector = count_vectorize_model.fit_transform([sentence_list[5]]).toarray()

word_frequency = pd.DataFrame(data = n_grams_feature_vector, columns = count_vectorize_model.get_feature_names_out())

print('Example sentence:', sentence_list[5])
word_frequency

Example sentence: giảng viên đảm bảo thời gian lên lớp , tích cực trả lời câu hỏi của sinh viên , thường xuyên đặt câu hỏi cho sinh viên .


bảo  cho  câu  của  cực  gian  giảng  hỏi  lên  lớp  lời  sinh  thường  \
0    1    1    2    1    1     1      1    2    1    1    1     2       1   

   thời  trả  tích  viên  xuyên  đảm  đặt  
0     1    1     1     3      1    1    1

In [16]:
count_vectorize_model = CountVectorizer(ngram_range = (1, 2))

n_grams_feature_vector = count_vectorize_model.fit_transform(example_sentence).toarray()

word_frequency = pd.DataFrame(data = n_grams_feature_vector, columns = count_vectorize_model.get_feature_names_out())
word_frequency.T

0  1  2  3
and              0  0  1  0
and this         0  0  1  0
document         1  2  0  1
document is      0  1  0  0
first            1  0  0  1
first document   1  0  0  1
is               1  1  1  1
is the           1  1  1  0
is this          0  0  0  1
one              0  0  1  0
second           0  1  0  0
second document  0  1  0  0
the              1  1  1  1
the first        1  0  0  1
the second       0  1  0  0
the third        0  0  1  0
third            0  0  1  0
third one        0  0  1  0
this             1  1  1  1
this document    0  1  0  0
this is          1  0  1  0
this the         0  0  0  1

## Problem set 1
Based on the UIT-VSFC dataset and the aforementioned information.
- Create an $n$-gram word frequency table, such that $n$ could be any number of your desire.
- With $n=1$ and $n=2$, what is the most popular word in the dataset ?
- With $n=1$ and $n=2$, what is the rarest word in the dataset ?
- What are the limitations of this data processing flow ? How can we overcome those ?


## Retrieve all sentences within the dataset

In [17]:
from typing import List

def get_all_sentences(dataset) -> List[str]:
    """
    Function to get all sentences and store them into a list of strings

    Args:
    dataset -- The subset (i.e., train/valid/test) in UIT-VSFC dataset

    Returns:
    A list of all sentences in a subset data of the UIT-VSFC.
    """

    list_all_sentence: list = []

    ### YOUR CODE STARTS HERE

    # Iterate through each example in the dataset and join words to form a sentence
    for idx in range(len(dataset)):
        sentence = ""
        # Join words in the sentence to form a full sentence
        for word in dataset[idx]['sentence']:
            sentence += word
            # Append the sentence to the list
        list_all_sentence.append(sentence)
    list_all_sentence

    ### YOUR CODE ENDS HERE


    return list_all_sentence

In [18]:
list_all_sentence: list = get_all_sentences(train_set)
print(f"#sentences within the dataset: {len(list_all_sentence)}")
print(f"Example sentence: {list_all_sentence[0]}")

#sentences within the dataset: 11426
Example sentence: slide giáo trình đầy đủ .


## Build the word frequency table

In [19]:
def n_gram_word_frequency(sentence_list: list, n: int) -> pd.DataFrame:
    """
    Function to build a word frequency table based on n-grams

    Args:
    sentence_list (list) -- A list of all sentences needed for table constructing process
    n (int) -- Number of grams that we parse into this function

    Returns:
    A dataframe contains all words after conducting n-grams and their respective frequencies
    """

    ### YOUR CODE STARTS HERE

    ## Create a CountVectorizer model with n-grams range
    count_vectorize_model = CountVectorizer(ngram_range = (n, n))
    # Fit the model to the sentence list and transform it into a feature vector
    n_grams_feature_vector = count_vectorize_model.fit_transform(sentence_list).toarray()
    # Create a DataFrame from the feature vector with appropriate column names
    word_frequency_table = pd.DataFrame(data = n_grams_feature_vector, columns = count_vectorize_model.get_feature_names_out())

    ### YOUR CODE ENDS HERE

    return word_frequency_table

In [20]:
# Construct the table of word frequency
word_frequency_table = n_gram_word_frequency(sentence_list=list_all_sentence,
                                             n=1 )
word_frequency_table

10  100  10h  10h30  11  11doubledot55  11h30  11h55  12  \
0       0    0    0      0   0              0      0      0   0   
1       0    0    0      0   0              0      0      0   0   
2       0    0    0      0   0              0      0      0   0   
3       0    0    0      0   0              0      0      0   0   
4       0    0    0      0   0              0      0      0   0   
...    ..  ...  ...    ...  ..            ...    ...    ...  ..   
11421   0    0    0      0   0              0      0      0   0   
11422   0    0    0      0   0              0      0      0   0   
11423   0    0    0      0   0              0      0      0   0   
11424   0    0    0      0   0              0      0      0   0   
11425   0    0    0      0   0              0      0      0   0   

       12doubledot00  ...  ấy  ẩn  ắt  ốc  ồn  ổn  ủa  ủng  ức  ứng  
0                  0  ...   0   0   0   0   0   0   0    0   0    0  
1                  0  ...   0   0   0   0   0   0   0    0   0    0  
2                  0  ...   0   0   0   0   0   0   0    0   0    0  
3                  0  ...   0   0   0   0   0   0   0    0   0    0  
4                  0  ...   0   0   0   0   0   0   0    0   0    0  
...              ...  ...  ..  ..  ..  ..  ..  ..  ..  ...  ..  ...  
11421              0  ...   0   0   0   0   0   0   0    0   0    0  
11422              0  ...   0   0   0   0   0   0   0    0   0    0  
11423              0  ...   0   0   0   0   0   0   0    0   0    0  
11424              0  ...   0   0   0   0   0   0   0    0   0    0  
11425              0  ...   0   0   0   0   0   0   0    0   0    0  

[11426 rows x 2459 columns]

You should comment your answer to problem 1 here with sufficient explanations, including your implementation and reasoning.

##1(a) Create an $n$-gram word frequency table, such that $n$ could be any number of your desire.


In [21]:
n = int(input("Enter the number of grams you want to parse: "))
word_frequency_table = n_gram_word_frequency(sentence_list=list_all_sentence,
                                             n=n )

Enter the number of grams you want to parse: 6


In [22]:
word_frequency_table

10 bài tập có thể khiến  10 fraction 10 điểm cho thầy  \
0                            0                             0   
1                            0                             0   
2                            0                             0   
3                            0                             0   
4                            0                             0   
...                        ...                           ...   
11421                        0                             0   
11422                        0                             0   
11423                        0                             0   
11424                        0                             0   
11425                        0                             0   

       10 kiến thức thầy dạy thậm  10 mấy năm đi học cả  \
0                               0                     0   
1                               0                     0   
2                               0                     0   
3                               0                     0   
4                               0                     0   
...                           ...                   ...   
11421                           0                     0   
11422                           0                     0   
11423                           0                     0   
11424                           0                     0   
11425                           0                     0   

       10 người nhưng sau ngày thì  10 người nhưng thầy vẫn chưa  \
0                                0                             0   
1                                0                             0   
2                                0                             0   
3                                0                             0   
4                                0                             0   
...                            ...                           ...   
11421                            0                             0   
11422                            0                             0   
11423                            0                             0   
11424                            0                             0   
11425                            0                             0   

       10 năm trước rất cũ so  10 sinh viên phải dùng laptop  \
0                           0                              0   
1                           0                              0   
2                           0                              0   
3                           0                              0   
4                           0                              0   
...                       ...                            ...   
11421                       0                              0   
11422                       0                              0   
11423                       0                              0   
11424                       0                              0   
11425                       0                              0   

       10 thì điểm thực hành cũng  10 trong lớp nên đưa ra  ...  \
0                               0                        0  ...   
1                               0                        0  ...   
2                               0                        0  ...   
3                               0                        0  ...   
4                               0                        0  ...   
...                           ...                      ...  ...   
11421                           0                        0  ...   
11422                           0                        0  ...   
11423                           0                        0  ...   
11424                           0                        0  ...   
11425                           0                        0  ...   

       ứng được thì tại sao trường  ứng được tiêu chí chấm điểm  \
0                                0                            0   
1   

## With  n=1  and  n=2 , what is the most popular word in the dataset ?

In [46]:
unigram = n_gram_word_frequency(sentence_list=list_all_sentence,
                                             n=1 )
bigram = n_gram_word_frequency(sentence_list=list_all_sentence,
                                             n=2 )

# Find the most frequent unigram
unigram_most_frequent = unigram.sum().idxmax()
unigram_frequency = unigram.sum().max()
print(f"Most frequent unigram: {unigram_most_frequent} (Frequency: {unigram_frequency})")

# Find the most frequent bigram
bigram_most_frequent = bigram.sum().idxmax()
bigram_frequency = bigram.sum().max()
print(f"Most frequent bigram: {bigram_most_frequent} (Frequency: {bigram_frequency})")


Most frequent unigram: viên (Frequency: 4803)
Most frequent bigram: sinh viên (Frequency: 2698)


##1(b) With  n=1  and  n=2 , what is the rarest word in the dataset ?

In [47]:
# Find all rarest unigrams
unigram_rarest_list = unigram.columns[unigram.sum() == unigram.sum().min()].tolist()
unigram_rarest_frequency = unigram.sum().min()
print(f"(Frequency: {unigram_rarest_frequency}) Rarest unigrams: {unigram_rarest_list} ")

# Find all rarest bigrams
bigram_rarest_list = bigram.columns[bigram.sum() == bigram.sum().min()].tolist()
bigram_rarest_frequency = bigram.sum().min()
print(f"(Frequency: {bigram_rarest_frequency}) Rarest bigrams: {bigram_rarest_list} ")


(Frequency: 1) Rarest unigrams: ['10h', '10h30', '11', '11doubledot55', '11h30', '11h55', '12doubledot00', '13', '13h45', '19', '1983', '1h30', '2007', '2011', '2015', '2048', '24', '245', '27', '2k3', '320k', '3dsmax', '3h15', '47', '49', '5h', '67', '70', '7h', '7h15', '7h20', '85', '92', '95', '9dot5', '9h45', 'a2', 'abc', 'add', 'aep', 'agent', 'agile', 'altera', 'amp', 'an', 'and', 'app', 'are', 'assembly', 'assignments', 'at', 'av1', 'av2', 'av3', 'bayes', 'best', 'bi', 'bia', 'bohm', 'book', 'boss', 'box', 'bridge', 'bus', 'buôi', 'buối', 'buốn', 'by', 'bào', 'bàu', 'bán', 'bâng', 'bê', 'bôi', 'búa', 'băm', 'băn', 'bơi', 'bưng', 'bạnem', 'bấy', 'bắc', 'bắng', 'bẻ', 'bốc', 'bồ', 'bổn', 'bụt', 'bự', 'bực', 'career', 'case', 'cau', 'cdio', 'cha', 'chair', 'chay', 'check', 'chip', 'chui', 'chuyến', 'chuông', 'chuốt', 'chàn', 'chìa', 'chót', 'chùi', 'chường', 'chạm', 'chảy', 'chấn', 'chấy', 'chập', 'chắp', 'chắt', 'chẵn', 'chặn', 'chối', 'chốn', 'chốt', 'chồng', 'chội', 'chờn', 'clas

##1(c) What are the limitations of this data processing flow? How can we overcome those?

Limitations of the current data processing flow
- Limitations of N-grams
1. Data Sparsity

Example: If n=3 (trigrams), many word combinations may appear very rarely, making the model unreliable.

Solution: Use backoff and smoothing techniques (like Laplace smoothing) to handle missing n-grams.

2. Out-of-Vocabulary (OOV) Words

Example: If a test dataset contains new words (e.g., "COVID-19" in 2020), an n-gram model trained before 2020 won’t recognize it.

Solution: Use subword tokenization (like Byte Pair Encoding - BPE) to break words into smaller meaningful parts.

3. Contextual Limitation

Example: "Apple" in "I ate an apple" vs. "Apple released a new iPhone" are treated as identical.

Solution: Use word embeddings (Word2Vec, GloVe, or BERT) instead of n-grams to capture context.

4. Long-Distance Dependencies

Example: In "The cat, which was very hungry, ate the fish," an n-gram model might fail to connect "cat" and "ate" if n is too small.

Solution: Use transformer-based models (BERT, GPT) which capture long-term dependencies.

- Limitations of TF-IDF
1. Inability to Capture Synonyms

Example: "happy" and "joyful" are treated as completely different words.

Solution: Use word embeddings (Word2Vec) to capture semantic similarity.

2. Ignoring Word Order

Example: "not good" and "good not" would have the same TF-IDF score, even though they mean different things.

Solution: Use n-grams in combination with TF-IDF to capture some order information.

3. Bias Towards Long Documents

Example: A long document will naturally have higher TF values, making it seem more important than a shorter but more relevant one.

Solution: Use normalization techniques like BM25 instead of plain TF-IDF.

Overcoming Limitations  
- **Dimensionality reduction techniques**: Techniques like Principal Component Analysis (PCA) or Singular Value Decomposition (SVD) can be used to reduce the dimensionality of the feature space and address sparsity.  
- **Smoothing techniques**: Techniques like Laplace smoothing or Add-k smoothing can be used to handle OOV words and prevent zero counts in the frequency table.  
- **Consider using higher-order n-grams**: In some cases, using higher-order n-grams (e.g., trigrams or 4-grams) may capture more contextual information, but this can also worsen sparsity issues.  

N-grams  
- **Dimensionality reduction techniques**: Techniques like Principal Component Analysis (PCA) or Singular Value Decomposition (SVD) can be used to reduce the dimensionality of the feature space and address sparsity.  
- **Smoothing techniques**: Techniques like Laplace smoothing or Add-k smoothing can be used to handle OOV words and prevent zero counts in the frequency table.  
- **Consider using higher-order n-grams**: In some cases, using higher-order n-grams (e.g., trigrams or 4-grams) may capture more contextual information, but this can also worsen sparsity issues.  

TF-IDF  
- **Word embeddings**: Word embeddings like Word2Vec or GloVe can be used to capture semantic relationships between words and provide a more context-aware representation of text.  
- **Advanced language models**: Pre-trained language models like BERT or RoBERTa can be used to extract more nuanced and contextualized features from text.  
- **Normalization techniques**: Different normalization techniques can be applied to TF-IDF to address bias towards longer documents.  
- **Combining with other features**: TF-IDF can be combined with other features like sentiment scores or part-of-speech tags to enhance the representation of text.  

General Improvements  
- **Data cleaning and preprocessing**: Proper data cleaning and preprocessing, including removing stop words, punctuation, and performing stemming or lemmatization, can significantly improve the quality of features extracted.  
- **Domain-specific knowledge**: Incorporating domain-specific knowledge and lexicons can help improve the accuracy and relevance of text processing.  
- **Experimentation**: It's crucial to experiment with different techniques and parameters to find the best approach for the specific dataset and task.  

By addressing these limitations and incorporating these improvements, we can build more robust and effective text processing pipelines for natural language processing tasks.

## Stopwords

In [25]:
# Retrieve the stopword dictionary
!wget --no-check-certificate --content-disposition https://raw.githubusercontent.com/stopwords/vietnamese-stopwords/master/vietnamese-stopwords.txt

--2025-03-28 16:08:49--  https://raw.githubusercontent.com/stopwords/vietnamese-stopwords/master/vietnamese-stopwords.txt
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.108.133, 185.199.109.133, 185.199.110.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.108.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 20475 (20K) [text/plain]
Saving to: ‘vietnamese-stopwords.txt’

vietnamese-stopword 100%[===================>]  20.00K  --.-KB/s    in 0.003s  

2025-03-28 16:08:49 (7.03 MB/s) - ‘vietnamese-stopwords.txt’ saved [20475/20475]



In [26]:
# Observe stopwords list
with open('vietnamese-stopwords.txt', 'r', encoding='utf-8') as file:
    vietnamese_stopword = file.read()

vietnamese_stopword = vietnamese_stopword.split('\n')  # Separate lines by lines
print(f"#Number of stop words: {len(vietnamese_stopword)}")


#Number of stop words: 1942


In [27]:
# Stop words example
for sentence in vietnamese_stopword[:10]:
    print(sentence)

a lô
a ha
ai
ai ai
ai nấy
ai đó
alô
amen
anh
anh ấy


## Term frequency - Invert document frequency (TF-IDF)


### TF
Term frequency (TF) is the number of times a given term appears in document

$$
tf(t) = f(t,d)\times\frac{1}{T}
$$
whereas, $f(t,d)$ is the frequency of the word $t$ in the document $d$, $T$ is the number of all words in that document.

In [28]:
import pandas as pd
from sklearn.feature_extraction.text import TfidfVectorizer

corpus = [
    'This is the first document.',
    'This document is the second document.',
    'And this is the third one.',
    'Is this the first document?',
]

# Declare TF vectorize
tf_vectorizer = TfidfVectorizer(ngram_range=(1, 1),
                                use_idf=False, # only using TF
                                norm='l1')

tf_vectorizer.fit_transform(corpus)

tf_vectorized = tf_vectorizer.transform(corpus)

tf_output = tf_vectorized[0]

# Build TF table
words_tf_idf = pd.DataFrame(tf_output.T.todense(), index=tf_vectorizer.get_feature_names_out(), columns=['tf'])
words_tf_idf

tf
and       0.0
document  0.2
first     0.2
is        0.2
one       0.0
second    0.0
the       0.2
third     0.0
this      0.2

### IDF

Inverse Document Frequency, or abbreviated as IDF, measures how important a term is. While computing TF, all terms are considered equally important. However it is known that certain terms, such as "is", "of", and "that", may appear a lot of times but have little importance. Thus we need to weigh down the frequent terms while scale up the rare ones.

$$
idf(t) = \log\left(\frac{\text{#documents in the document set}}{\text{#documents with term}}\right) + 1
$$

In [29]:
import pandas as pd
from sklearn.feature_extraction.text import TfidfVectorizer

corpus = [
    'This is the first document.',
    'This document is the second document.',
    'And this is the third one.',
    'Is this the first document?',
]

# Configure settings for IDF vectorize
tf_idf_vectorizer = TfidfVectorizer(ngram_range=(1, 1),
                                    smooth_idf=False,
                                    use_idf=True,
                                    norm=None)

tf_idf_vectorizer.fit_transform(corpus)

# Retrieve only idf information
idf_vectorizer = tf_idf_vectorizer.idf_

# Join idf values into the previous dataframe
words_tf_idf['idf'] = idf_vectorizer

# Show dataframe with ascending values of idf
words_tf_idf.sort_values(by=['idf'])

tf       idf
is        0.2  1.000000
the       0.2  1.000000
this      0.2  1.000000
document  0.2  1.287682
first     0.2  1.693147
and       0.0  2.386294
second    0.0  2.386294
one       0.0  2.386294
third     0.0  2.386294

### TF-IDF

Technically saying, TF-IDF is a score which is applied to every word in every document in our dataset. And for every word, the TF-IDF value increases with every appearance of the word in a document, but is gradually decreased with every appearance in other documents

$$
\text{tf-idf}= tf(t, d) \times idf(t)
$$

In [30]:
import pandas as pd
from sklearn.feature_extraction.text import TfidfVectorizer

corpus = [
    'This is the first document.',
    'This document is the second document.',
    'And this is the third one.',
    'Is this the first document?',
]

tf_idf_vectorizer = TfidfVectorizer(ngram_range=(1, 1),
                                    smooth_idf=False,
                                    use_idf=True,
                                    norm='l1')

tf_idf_vectorizer.fit_transform(corpus)

tf_idf_vectorized = tf_idf_vectorizer.transform(corpus)

tf_idf_output = tf_idf_vectorized[0]
words_tf_idf['tf-idf'] = tf_idf_output.T.todense()

words_tf_idf.sort_values(by=['tf-idf'])

tf       idf    tf-idf
and       0.0  2.386294  0.000000
third     0.0  2.386294  0.000000
second    0.0  2.386294  0.000000
one       0.0  2.386294  0.000000
is        0.2  1.000000  0.167201
the       0.2  1.000000  0.167201
this      0.2  1.000000  0.167201
document  0.2  1.287682  0.215302
first     0.2  1.693147  0.283096

### Problem set 2
Based on the problem 1 and the instruction on TF, IDF, TF-IDF:
- (2a) Build the tf-idf table for the UIT-VSFC dataset with $n$-gram = 1 and $n$-gram = 2.
- (2b) Change a few hyperparameters in the `TfidfVectorizer` function (`smooth_idf`, `sublinear_tf` and `norm`) from problem 2a (*you could browse from this [link](https://scikit-learn.org/stable/modules/generated/sklearn.feature_extraction.text.TfidfVectorizer.html) to discover which are the correct paramters to parse*). Explain the results differences collected after modifying hyperparameters.
- (2c) Which words has the lowest and the highest tf-idf values ? Do they differ from $n$-grams results ?
- (2d) Which limitations from $n$-grams that TF-IDF overcame ?

##  (2a) Build the tf-idf table for the UIT-VSFC dataset with $n$-gram = 1 and $n$-gram = 2.

In [31]:
def compute_tf_idf(corpus: list, n=1, smooth_idf=False, sublinear_tf=False, norm='l2') -> pd.DataFrame:
    """
    Compute the TF-IDF table for a given corpus.

    Args:
        corpus (list): List of sentences/documents.
        n (int): The n-gram value to consider (default is unigram).
        smooth_idf (bool): Whether to smooth IDF weights by adding 1 to document frequencies (default is False).
        norm (str): Normalization to apply to term vectors (default is None).
        sublinear_tf (bool): Whether to use sublinear term frequency scaling (default is False).

    Returns:
        pd.DataFrame: A DataFrame containing the TF-IDF matrix.
    """

    # Configure TfidfVectorizer with the given parameters
    vectorizer = TfidfVectorizer(
        ngram_range=(n, n),
        smooth_idf=smooth_idf,
        use_idf=True,
        norm=norm,
        sublinear_tf=sublinear_tf
    )

    # Fit and transform the corpus to get TF-IDF matrix directly
    tf_idf_matrix = vectorizer.fit_transform(corpus)

    # A
    idf_values = vectorizer.idf_

    # Convert the TF-IDF matrix to a dense array (if needed) and then to DataFrame
    tf_matrix = tf_idf_matrix.copy()  # Here tf_idf_matrix already has the effect of TF*IDF


    # Convert the matrix to a dense array  and then to DataFrame
    df = pd.DataFrame(tf_idf_matrix.toarray(), columns=vectorizer.get_feature_names_out())
    return df


In [32]:
# Compute TF-IDF for unigrams (n=1)
tfidf_unigram = compute_tf_idf(corpus=list_all_sentence, n=1)
print("TF-IDF for unigrams:")
print(tfidf_unigram)

# Compute TF-IDF for bigrams (n=2)
tfidf_bigram = compute_tf_idf(corpus=list_all_sentence, n=2)
print("\nTF-IDF for bigrams:")
print(tfidf_bigram)


TF-IDF for unigrams:
        10  100  10h  10h30   11  11doubledot55  11h30  11h55   12  \
0      0.0  0.0  0.0    0.0  0.0            0.0    0.0    0.0  0.0   
1      0.0  0.0  0.0    0.0  0.0            0.0    0.0    0.0  0.0   
2      0.0  0.0  0.0    0.0  0.0            0.0    0.0    0.0  0.0   
3      0.0  0.0  0.0    0.0  0.0            0.0    0.0    0.0  0.0   
4      0.0  0.0  0.0    0.0  0.0            0.0    0.0    0.0  0.0   
...    ...  ...  ...    ...  ...            ...    ...    ...  ...   
11421  0.0  0.0  0.0    0.0  0.0            0.0    0.0    0.0  0.0   
11422  0.0  0.0  0.0    0.0  0.0            0.0    0.0    0.0  0.0   
11423  0.0  0.0  0.0    0.0  0.0            0.0    0.0    0.0  0.0   
11424  0.0  0.0  0.0    0.0  0.0            0.0    0.0    0.0  0.0   
11425  0.0  0.0  0.0    0.0  0.0            0.0    0.0    0.0  0.0   

       12doubledot00  ...   ấy   ẩn   ắt   ốc   ồn   ổn   ủa  ủng   ức  ứng  
0                0.0  ...  0.0  0.0  0.0  0.0  0.0  0.0  0.0

#### (2b) Change a few hyperparameters in the `TfidfVectorizer` function (`smooth_idf`, `sublinear_tf` and `norm`) from problem 2a (*you could browse from this [link] to discover which are the correct paramters to parse). Explain the results differences collected after modifying hyperparameters.

In [33]:
# Test different hyperparameter combinations with the first document
smooth_idf_values = [True, False]
sublinear_tf_values = [True, False]
norm_values = ['l1','l2']

# Iterate through all combinations of hyperparameters with unigram n-gram
for smooth_idf in smooth_idf_values:
    for sublinear_tf in sublinear_tf_values:
        for norm in norm_values:
            print(f"Testing with smooth_idf={smooth_idf}, sublinear_tf={sublinear_tf}, norm={norm}")
            tfidf_table = compute_tf_idf(corpus=list_all_sentence, n=1, smooth_idf=smooth_idf, sublinear_tf=sublinear_tf, norm=norm)

            # Print the first row, transpose it, and sort by TF-IDF values in descending order
            print(tfidf_table.iloc[0:1].T.sort_values(by=0, ascending=False))

            print("\n" + "="*50 + "\n")


Testing with smooth_idf=True, sublinear_tf=True, norm=l1
               0
slide   0.217528
đầy     0.216997
đủ      0.202504
trình   0.184496
giáo    0.178474
...          ...
khoát   0.000000
khoăn   0.000000
khoảng  0.000000
khu     0.000000
khiến   0.000000

[2459 rows x 1 columns]


Testing with smooth_idf=True, sublinear_tf=True, norm=l2
               0
slide   0.484828
đầy     0.483643
đủ      0.451342
trình   0.411205
giáo    0.397784
...          ...
khoát   0.000000
khoăn   0.000000
khoảng  0.000000
khu     0.000000
khiến   0.000000

[2459 rows x 1 columns]


Testing with smooth_idf=True, sublinear_tf=False, norm=l1
               0
slide   0.217528
đầy     0.216997
đủ      0.202504
trình   0.184496
giáo    0.178474
...          ...
khoát   0.000000
khoăn   0.000000
khoảng  0.000000
khu     0.000000
khiến   0.000000

[2459 rows x 1 columns]


Testing with smooth_idf=True, sublinear_tf=False, norm=l2
               0
slide   0.484828
đầy     0.483643
đủ      0.451342
trình   0

Static metric n = 1

In [37]:
# Danh sách các bộ tham số
smooth_idf_values = [True, False]
sublinear_tf_values = [True, False]
norm_values = ['l1', 'l2']

# Duyệt qua từng bộ tham số
results = []
for smooth_idf in smooth_idf_values:
    for sublinear_tf in sublinear_tf_values:
        for norm in norm_values:
            print(f"Testing with smooth_idf={smooth_idf}, sublinear_tf={sublinear_tf}, norm={norm}")
            tfidf_table = compute_tf_idf(corpus=list_all_sentence, n=1, smooth_idf=smooth_idf, sublinear_tf=sublinear_tf, norm=norm)

            # Tính toán thống kê
            mean_value = tfidf_table.mean().mean()
            min_value = tfidf_table.min().min()
            max_value = tfidf_table.max().max()
            std = tfidf_table.std().std()

            # Lưu kết quả
            results.append({
                "smooth_idf": smooth_idf,
                "sublinear_tf": sublinear_tf,
                "norm": norm,
                "mean": mean_value,
                "min": min_value,
                "max": max_value,
                "std" : std
            })




Testing with smooth_idf=True, sublinear_tf=True, norm=l1
Testing with smooth_idf=True, sublinear_tf=True, norm=l2
Testing with smooth_idf=True, sublinear_tf=False, norm=l1
Testing with smooth_idf=True, sublinear_tf=False, norm=l2
Testing with smooth_idf=False, sublinear_tf=True, norm=l1
Testing with smooth_idf=False, sublinear_tf=True, norm=l2
Testing with smooth_idf=False, sublinear_tf=False, norm=l1
Testing with smooth_idf=False, sublinear_tf=False, norm=l2


In [38]:
# change data frame
static_results = pd.DataFrame(results)
print(static_results)

   smooth_idf  sublinear_tf norm      mean  min  max       std
0        True          True   l1  0.000407  0.0  1.0  0.005736
1        True          True   l2  0.001266  0.0  1.0  0.014830
2        True         False   l1  0.000407  0.0  1.0  0.005758
3        True         False   l2  0.001260  0.0  1.0  0.014898
4       False          True   l1  0.000407  0.0  1.0  0.005717
5       False          True   l2  0.001265  0.0  1.0  0.014752
6       False         False   l1  0.000407  0.0  1.0  0.005739
7       False         False   l2  0.001259  0.0  1.0  0.014821


1. Effect of Normalization (norm)
L1 normalization (l1) results in significantly lower mean values (0.000407) compared to L2 normalization (l2) (0.00126).

L2 normalization (l2) results in higher standard deviation (std) values (~0.0148) than L1 normalization (l1) (0.0057), meaning L2 distributes values more widely.

L1 focuses on sparsity, while L2 smooths out the impact of different terms, leading to greater variance in TF-IDF scores.

2. Effect of smooth_idf (IDF Smoothing)
Smoothing (smooth_idf=True) has a negligible impact on the mean and standard deviation values.

The difference in std when comparing smooth_idf=True vs False is minimal (~0.00002-0.0001).

This suggests that IDF smoothing does not significantly affect the distribution of TF-IDF scores but helps in avoiding division by zero issues.

3. Effect of sublinear_tf (Logarithmic Scaling of Term Frequency)
With sublinear_tf=True, the standard deviation (std) tends to be slightly lower for l2 normalization.

Log-scaling (sublinear_tf=True) has a more noticeable impact on l2 normalization, making values slightly more stable (std ~0.0148 vs. ~0.0149).

For l1 normalization, log-scaling has almost no effect (mean ≈ 0.000407).

4. Maximum (max=1.0) Across All Cases
The maximum TF-IDF score remains 1.0 for all parameter configurations.

This suggests that at least one term in each case has the highest possible TF-IDF score, likely due to normalization.

similar with n = 2


In [39]:
# Danh sách các bộ tham số
smooth_idf_values = [True, False]
sublinear_tf_values = [True, False]
norm_values = ['l1', 'l2']

# Duyệt qua từng bộ tham số
results2 = []
for smooth_idf in smooth_idf_values:
    for sublinear_tf in sublinear_tf_values:
        for norm in norm_values:
            print(f"Testing with smooth_idf={smooth_idf}, sublinear_tf={sublinear_tf}, norm={norm}")
            tfidf_table = compute_tf_idf(corpus=list_all_sentence, n=2, smooth_idf=smooth_idf, sublinear_tf=sublinear_tf, norm=norm)

            # Tính toán thống kê
            mean_value = tfidf_table.mean().mean()
            min_value = tfidf_table.min().min()
            max_value = tfidf_table.max().max()
            std = tfidf_table.std().std()

            # Lưu kết quả
            results2.append({
                "smooth_idf": smooth_idf,
                "sublinear_tf": sublinear_tf,
                "norm": norm,
                "mean": mean_value,
                "min": min_value,
                "max": max_value,
                "std" : std
            })




Testing with smooth_idf=True, sublinear_tf=True, norm=l1
Testing with smooth_idf=True, sublinear_tf=True, norm=l2
Testing with smooth_idf=True, sublinear_tf=False, norm=l1
Testing with smooth_idf=True, sublinear_tf=False, norm=l2
Testing with smooth_idf=False, sublinear_tf=True, norm=l1
Testing with smooth_idf=False, sublinear_tf=True, norm=l2
Testing with smooth_idf=False, sublinear_tf=False, norm=l1
Testing with smooth_idf=False, sublinear_tf=False, norm=l2


In [40]:
# change data frame
static_results2 = pd.DataFrame(results2)
print(static_results2)

   smooth_idf  sublinear_tf norm      mean  min  max       std
0        True          True   l1  0.000407  0.0  1.0  0.005736
1        True          True   l2  0.001266  0.0  1.0  0.014830
2        True         False   l1  0.000407  0.0  1.0  0.005758
3        True         False   l2  0.001260  0.0  1.0  0.014898
4       False          True   l1  0.000407  0.0  1.0  0.005717
5       False          True   l2  0.001265  0.0  1.0  0.014752
6       False         False   l1  0.000407  0.0  1.0  0.005739
7       False         False   l2  0.001259  0.0  1.0  0.014821


TF-IDF Hyperparameters
======================================

When processing text with TF-IDF (Term Frequency-Inverse Document Frequency), several hyperparameters can be adjusted. These include **smooth_idf**, **sublinear_tf**, and **norm**. The following explanation describes each parameter with simple language and examples suitable for a high school level, without using second-person pronouns.

Smooth_idf
----------
**Functionality:**  
This parameter adds a small constant to the count of documents in which a term appears. It prevents rare terms from obtaining an excessively high weight.

**Analogy:**  
Imagine grading homework submissions.  
- Without smoothing: A student who submits homework only once might receive an unusually high score for uniqueness.  
- With smoothing (`smooth_idf=True`): A small bonus is added to every student's score, ensuring that a one-time submission does not stand out excessively.

Sublinear_tf
------------
**Functionality:**  
Instead of using the raw frequency count of a term, this parameter applies a logarithmic scaling. This reduces the impact of very frequent terms.

**Analogy:**  
Consider counting candies in jars.  
- Without sublinear scaling: A jar containing 100 candies appears to be 10 times larger than a jar with 10 candies.  
- With sublinear scaling (`sublinear_tf=True`): The difference is less dramatic, balancing the influence of jars with many candies compared to those with fewer candies.

Norm
----
**Functionality:**  
Normalization scales the TF-IDF scores so that document vectors are comparable, regardless of the number of words or document length.

**Analogy:**  
Think of comparing test scores from two classes—one class with tests out of 100 points and another with tests out of 50 points.  
- **With normalization (e.g., 'l1' or 'l2'):** Scores are converted to a common scale, similar to expressing them as percentages.  
- **Without normalization (None):** Documents with more words or higher term frequencies may have inflated scores, which does not necessarily reflect better performance.

Summary
-------

- **Smooth IDF:**  
  - *Enabled:* Rare terms receive a more balanced weight.  
  - *Disabled:* Rare terms might obtain disproportionately high scores.
- **Sublinear TF:**  
  - *Enabled:* The influence of very frequent terms is moderated.  
  - *Disabled:* Frequent terms can dominate the scoring.
- **Normalization (Norm):**  
  - *Applied:* Document vectors are scaled for fair comparison.  
  - *Not applied:* Longer documents or those with repetitive content may exhibit inflated scores.


## (2c) Which words has the lowest and the highest tf-idf values ? Do they differ from  n -grams results ?

In [ ]:
# Compute TF-IDF for unigrams (n=1)
tfidf_unigram = compute_tf_idf(corpus=list_all_sentence, n=1)

min_tfidf_unigram = tfidf_unigram.min()
max_tfidf_unigram = tfidf_unigram.max()


In [ ]:
# list of the lowest tf-idf values
min_tfidf_unigram

10       0.0
100      0.0
10h      0.0
10h30    0.0
11       0.0
        ... 
ổn       0.0
ủa       0.0
ủng      0.0
ức       0.0
ứng      0.0
Length: 2459, dtype: float64

In [ ]:
# list of the highest tf-idf values
max_tfidf_unigram.sort_values(ascending=False)

dễ       1.000000
không    1.000000
hết      1.000000
hay      1.000000
hiền     1.000000
           ...   
tá       0.133096
gạo      0.127809
lõng     0.127809
ùn       0.127809
vạ       0.127809
Length: 2459, dtype: float64

In [ ]:
# Compute TF-IDF for unigrams (n=2)
tfidf_bigram = compute_tf_idf(corpus=list_all_sentence, n=2)

min_tfidf_bigram = tfidf_bigram.min()
max_tfidf_bigram = tfidf_bigram.max()


In [ ]:
min_tfidf_bigram.sort_values(ascending=True)

ứng nhu     0.0
ứng kịp     0.0
ứng dụng    0.0
ứng cho     0.0
ức chế      0.0
           ... 
ứng đúng    0.0
ứng được    0.0
ứng đầy     0.0
ứng đủ      0.0
10 50       0.0
Length: 31384, dtype: float64

In [ ]:
max_tfidf_bigram.sort_values(ascending=False)

giảng viên    1.000000
quan tâm      1.000000
giảng dạy     1.000000
rất nản       1.000000
cám ơn        1.000000
                ...   
quá nhiểu     0.098609
lõng mơ       0.098609
tính như      0.098609
như chúng     0.098609
như phần      0.098609
Length: 31384, dtype: float64

In [53]:
def get_tf_idf_extremes(df):
    """
    Find the word with the lowest and highest TF-IDF values, and print the list of words with the minimum TF-IDF value.

    Args:
        df (pd.DataFrame): DataFrame containing TF-IDF values for words.

    Returns:
        tuple: A tuple containing:
            - max_word (str): Word with the highest TF-IDF value.
            - min_value (float): The lowest TF-IDF value.
            - max_value (float): The highest TF-IDF value.
            - min_words_list (list): List of words with the lowest TF-IDF value.
    """
    # Word with lowest TF-IDF
    min_value = df.min().min()
    min_words_list = df.columns[df.min() == min_value].tolist()

    # Word with highest TF-IDF
    max_word = df.max().idxmax()
    max_value = df.max().max()

    return max_word, min_value, max_value, min_words_list

the highest tfidf value and the lowest ifidf values with default parameter


In [55]:
max_word_unigram, min_value_unigram, max_value_unigram, min_words_list_unigram = get_tf_idf_extremes(tfidf_unigram)
max_word_bigram, min_value_bigram, max_value_bigram, min_words_list_bigram = get_tf_idf_extremes(tfidf_bigram)

print(f" Highest TF-IDF word: {max_word_unigram} (Value: {max_value_unigram}) Unigrams: Lowest TF-IDF word(s): {min_words_list_unigram} (Value: {min_value_unigram}), ")
print(f" Highest TF-IDF word: {max_word_bigram} (Value: {max_value_bigram}) Bigrams: Lowest TF-IDF word(s): {min_words_list_bigram} (Value: {min_value_bigram}),")

# Print the list of lowest TF-IDF words for both unigrams and bigrams
print("\nDetailed list of lowest TF-IDF words:")
print(f"Unigrams with lowest TF-IDF: {min_words_list_unigram}")
print(f"Bigrams with lowest TF-IDF: {min_words_list_bigram}")


 Highest TF-IDF word: chán (Value: 1.0) Unigrams: Lowest TF-IDF word(s): ['10', '100', '10h', '10h30', '11', '11doubledot55', '11h30', '11h55', '12', '12doubledot00', '12h', '13', '13h30', '13h45', '15', '19', '1983', '1dot5', '1h', '1h30', '20', '2003', '2005', '2007', '2008', '2009', '2010', '2011', '2012', '2013', '2015', '2048', '23h59', '24', '245', '25', '27', '2h', '2k3', '30', '32', '320k', '3dsmax', '3h15', '3h30', '40', '45', '47', '49', '50', '5h', '60', '64', '67', '70', '75', '7h', '7h15', '7h20', '7h30', '80', '85', '8h', '90', '92', '95', '9dot5', '9h', '9h45', 'a2', 'abc', 'add', 'aep', 'agent', 'agile', 'ai', 'altera', 'am', 'amp', 'an', 'and', 'android', 'anh', 'app', 'are', 'assembly', 'assignments', 'at', 'av1', 'av2', 'av3', 'ba', 'bai', 'ban', 'bao', 'basic', 'bayes', 'best', 'bi', 'bia', 'bit', 'biên', 'biến', 'biếng', 'biết', 'biểu', 'biện', 'biệt', 'bohm', 'book', 'boss', 'box', 'bridge', 'bus', 'buôi', 'buối', 'buốn', 'buồn', 'buổi', 'buộc', 'by', 'bài', 'bàn'

Compare with n gram result there are different

At their core, TF-IDF and n-grams serve different purposes:
-----------------------------------
- TF-IDF is a weighting method applied to terms to assign importance based on frequency and rarity across documents.

- N-grams are a representation method for extracting sequences of words from text.
However, they are often used together: n-grams (unigrams, bigrams, etc.) can be the "terms" to which TF-IDF is applied. For example, you might calculate TF-IDF scores for bigrams like "love dogs" instead of just unigrams like "dogs". This suggests they are complementary rather than entirely separate.
based on result:
TF-IDF on Unigrams vs. N-gram Frequencies: The TF-IDF scores differ from raw bigram counts TF-IDF incorporates document rarity (IDF), while raw frequencies do not.

# **2d**) Which limitations from $n$-grams that TF-IDF overcame ?


Sparsity:
-----------------------------------
How TF-IDF Helps: While TF-IDF doesn’t eliminate sparsity, it reduces its impact by assigning low weights to rare n-grams that appear in few documents (via low IDF scores). This shifts the focus to more informative n-grams, mitigating the problem of sparse, infrequently occurring sequences.

High Dimensionality:
------------------------------------
How TF-IDF Helps: TF-IDF itself doesn’t reduce the number of features, but it enables effective feature selection. By prioritizing n-grams with high TF-IDF scores, less informative ones can be filtered out, indirectly addressing the high-dimensionality issue when paired with such techniques.
Sensitivity to Noise:
-------------------------------------
How TF-IDF Helps: By focusing on term importance rather than just frequency, TF-IDF can be somewhat more robust to noise. For example, if "love dogs" and "adore dogs" both contain "dogs" as a key term, TF-IDF might still highlight "dogs" as important, even if the exact n-grams differ due to synonyms or variations.
Limitations TF-IDF Does Not Overcome
---------------------------
TF-IDF doesn’t fully resolve all n-gram limitations:

Lack of Semantic Understanding: TF-IDF still treats n-grams as discrete units and cannot recognize semantic similarities (e.g., "not good" vs. "bad").
Context Beyond N-grams: It doesn’t capture word order or long-range dependencies beyond the chosen n, as it relies on a bag-of-words-like approach.


Contextual understanding: While n-grams capture some context by considering word sequences, TF-IDF goes beyond simple co-occurrence.  By incorporating IDF, it downweights frequent words (like "the" or "a") that appear in many documents but carry less specific meaning.  This implicitly emphasizes terms that are more distinctive to a particular document or topic.

Dimensionality reduction:  N-grams, especially higher-order ones, lead to a very high-dimensional feature space.  TF-IDF itself doesn't reduce dimensionality, but it assigns lower weights to less important features, making the feature space more manageable when used with dimensionality reduction techniques (like PCA or SVD) for downstream applications.  It helps to focus on the most relevant features.

Handling of frequent but less informative words: N-grams treat all words equally based on their frequency. TF-IDF addresses the issue of frequent, uninformative words by down-weighting them with IDF.  N-grams might overrepresent these words, making it difficult to discern the true topic or meaning of the document.

In essence, TF-IDF enhances n-grams by adding a weighting scheme that distinguishes between important and unimportant words within a document relative to a collection of documents.  It uses the inverse document frequency component to improve upon the limitations of purely counting n-gram occurrences.
